ASoC: SOF: Patch set to fix TLV bytes put related issues

singalsu opened the pull request:
This pull request consists of three commits. 

- The first adds a defined limit of 1 Mbytes for TLV bytes data and two checks for it in the code

- The second commit fixes changes the return value from put control to be passed to caller.

- The third commit updates the put controls to use first a temporary buffer for sending IPC request. The scontrol cache is updated if the control was successful.

Edit: Re-re-re-updated this note.

singalsu commented:
This fixes issue #753. I made this RFC because there's also other way to fix this (prevent response copy if -EBUSY, and maybe other ways too). However since I felt that with process components generally, it's better to be safe with errors and better to avoid copying at all user data to PM cache before seeing that it was accepted.

plbossart commented on sound/soc/sof/control.c:
```diff
@@ -458,15 +475,24 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
+					    scontrol->cmd,
+					    true);
+
+	/* If IPC was successful copy new data into PM cache */
```
What is a 'PM cache'?
Also isn't the issue rather that the IPC return value was not handled properly? I don't see the added value of the memcpy?

singalsu replied:
The cdata points to persistent memory for controls that I call "PM cache". With this patch the persistent memory is updated only if the IPC was successful.

singalsu replied:
Sounds like using [RFC] was useful. I'll update the comments to something better. @ranj063 What do you think about this approach?

ranj063 replied:
@singalsu when will you get a -EBUSY from the DSP?

ranj063 replied:
and does it mean you cannot apply a new config with sof-eqctl while playback is in progress?

plbossart replied:
> Sounds like using [RFC] was useful. I'll update the comments to something better. @ranj063 What do you think about this approach?

yes, please. using words like 'cache corruption' will raise tons of questions, when what you are really doing is making sure the data passed by userspace is only cached by the kernel for future suspend/resume transitions if the initial IPC succeeds.

ranj063 commented:
@singalsu the easier fix for this problem is that we dont memcpy the reply message in tx_wait_done() is reply.error is < 0, isnt it?

cujomalainey commented on sound/soc/sof/control.c:
```diff
@@ -431,24 +432,40 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
+		return -ENOMEM;
+
+	/* Point scontrol temporarily to the new data for IPC send */
+	scontrol->control_data = new_cdata;
```
Should we roll this back on a  ABI failure? I see if the TX fails it is rolled back, but not for any other errors.

cujomalainey replied:
Actually we should definitely roll this back, otherwise `scontrol->control_data` points to freed memory

ranj063 replied:
@cujomalainey it looks like Seppo does assign cdata back to scontrol->control_data further down. Is that what you're talking about?

ranj063 replied:
Ahh I see what you mean. yes, error handling is incomplete here

singalsu replied:
@cujomalainey Oops, thanks for finding this!

singalsu commented:
There’s the problem that the message won’t be re-sent to DSP if PM would not be used if just dropping the copy in tx done (when EBUSY).

Edit - though I have’t confirmed this.

singalsu commented:
The EBUSY happens if sof-eqtl tries to reconfigure an EQ when it’s active.

Though that limitation can be later fixed (if new EQ needs same or lower amount of RAM).

ranj063 commented:
> There’s the problem that the message won’t be re-sent to DSP if PM would not be used if just dropping the copy in tx done (when EBUSY).

@singalsu i dint get this point. So if the IPC fails, we wont copy the reply message in tx_wait_done. 
And when you resume from runtime suspend next time, cdata would still contain the new config from userspace because you copied it in bytes_ext_put() and it will be sent to the DSP right?

singalsu commented:
I don't understand why sof-ci/jenkins/pr-checkpatch and continuous-integration/travis-ci/pr keep checking old version of this PR.

singalsu commented:
So to explain more what happens is that with this PR attempt to configure an EQ during playback results to proper fail. It can be seen from sof-eqctl output.

```
singalsu@piikkisika:~$ sof-eqctl -n 22 -s eq/eq_iir_bassboost.txt 
Control size is 304.
Applying configuration "eq/eq_iir_bassboost.txt" into device hw:0 control numid=22.
4607827,0,116,50331648,0,0,0,0,116,2,1,0,0,0,0,0,0,2,2,0,0,0,0,3227172081,2141520527,536653443,3221660410,536653443,0,16384,3260252783,2107733822,161646111,3961037800,172645501,4294967294,27910
Error: failed TLV write.

```

Successive playback happens with old configuration unless a new configuration is passed when pipeline is idle.

I'll try the other approach and talk about it next vs. this PR.

singalsu commented:
This is another way to address the issue. I'm having my self uncertain feeling about omitting memcpy with nonzero ret (or just with -EBUSY?). Can this change impact something else?

Also for this version I added propagation of possible IPC error to use space. I think omitting of e.g. -EBUSY (or report it as a warning of delay to apply) should be handled in user space application like sof-eqctl. 

```
commit 9a0ee42d4ce0f058d1805a85b0fc2b6e488870b4 (HEAD -> fix_fw_response_copy_when_error)
Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
Date:   Tue Apr 30 12:54:15 2019 +0300

    [RFC]ASoC: SOF: Do not copy failed response from DSP
    
    This patch prevents copy of failed IPC command return data to
    scontrol persist data in tx_wait_done(). It prevents failed IPC
    to be re-applied by PM restore.
    
    The change for snd_sof_bytes_ext_put() propagates to user space
    potential fail of IPC command. Without this the user space shows
    always success even for failed IPC.
    
    Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

diff --git a/sound/soc/sof/control.c b/sound/soc/sof/control.c
index 11762c4580f1..400f0089d44f 100644
--- a/sound/soc/sof/control.c
+++ b/sound/soc/sof/control.c
@@ -462,11 +462,11 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
        }
 
        /* notify DSP of byte control updates */
-       snd_sof_ipc_set_get_comp_data(sdev->ipc, scontrol,
-                                     SOF_IPC_COMP_SET_DATA,
-                                     SOF_CTRL_TYPE_DATA_SET,
-                                     scontrol->cmd,
-                                     true);
+       ret = snd_sof_ipc_set_get_comp_data(sdev->ipc, scontrol,
+                                           SOF_IPC_COMP_SET_DATA,
+                                           SOF_CTRL_TYPE_DATA_SET,
+                                           scontrol->cmd,
+                                           true);
 
        pm_runtime_mark_last_busy(sdev->dev);
        err = pm_runtime_put_autosuspend(sdev->dev);
diff --git a/sound/soc/sof/ipc.c b/sound/soc/sof/ipc.c
index ba1bb17a8d1e..c1e67906c934 100644
--- a/sound/soc/sof/ipc.c
+++ b/sound/soc/sof/ipc.c
@@ -213,7 +213,7 @@ static int tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *msg,
        } else {
                /* copy the data returned from DSP */
                ret = msg->reply_error;
-               if (msg->reply_size)
+               if (msg->reply_size && !ret)
                        memcpy(reply_data, msg->reply_data, msg->reply_size);
                if (ret < 0)
                        dev_err(sdev->dev, "error: ipc error for 0x%x size %zu\n",

```
The sof-eqctl if done during playback reports an error but since the new config remains in PM cache it is applied when next playback starts. The next command shows the fail from -EBUSY but when reading the coefficients back it can be seen that they were applied (as well as hearing the boost EQ effect)

```
singalsu@piikkisika:~$ sof-eqctl -n 22 -s eq/eq_iir_bassboost.txt
Control size is 304.
Applying configuration "eq/eq_iir_bassboost.txt" into device hw:0 control numid=22.
4607827,0,116,50331648,0,0,0,0,116,2,1,0,0,0,0,0,0,2,2,0,0,0,0,3227172081,2141520527,536653443,3221660410,536653443,0,16384,3260252783,2107733822,161646111,3961037800,172645501,4294967294,27910
Error: failed TLV write.
singalsu@piikkisika:~$ sof-eqctl -n 22
Control size is 304.
Retrieving configuration for device hw:0 control numid=22.
Success.
4607827,0,116,50356224,0,0,0,0,116,2,1,0,0,0,0,0,0,2,2,0,0,0,0,3227172081,2141520527,536653443,3221660410,536653443,0,16384,3260252783,2107733822,161646111,3961037800,172645501,4294967294,27910

```

I'll test next what happens with this approach with PM suspend disabled.

singalsu commented:
The latter approach has the problem that if I disable PM suspend the control that was sent during playback does not impact the next playback. Here's same example as above where reading back the coefficients return the old configuration and new configuration does not apply.

```
root@piikkisika:~# echo on > "/sys/bus/pci/devices/0000:00:0e.0/power/control"
root@piikkisika:~# sof-eqctl -n 22 -s eq/eq_iir_bassboost.txt
Control size is 304.
Applying configuration "eq/eq_iir_bassboost.txt" into device hw:0 control numid=22.
4607827,0,116,50331648,0,0,0,0,116,2,1,0,0,0,0,0,0,2,2,0,0,0,0,3227172081,2141520527,536653443,3221660410,536653443,0,16384,3260252783,2107733822,161646111,3961037800,172645501,4294967294,27910
Error: failed TLV write.
root@piikkisika:~# sof-eqctl -n 22
Control size is 304.
Retrieving configuration for device hw:0 control numid=22.
Success.
4607827,0,88,50356224,0,0,0,0,88,2,1,0,0,0,0,4294967295,4294967295,1,1,0,0,0,0,0,0,0,0,538145694,0,32690

```

So the behavior of the latter approach depends on PM suspend setting while the first approach does not have this issue. Using this approach would need re-send of control code add with suspend disabled to make it consistent. Also I'm leaning towards treating -EBUSY as error and have no impact to EQ configuration in that case (the user needs to redo configuration until success is reported). But this is somewhat a matter of taste.

The existing behavior with corrupt cached control is a clear  bug in my opinion so some fix the the first or second approach (or something else) is needed.

ranj063 commented:
@singalsu with runtime PM enabled, you mentioned that applying the coefficients during playback shows an error but the coefficients were applied when you get the values it shows that the new config was applied but there is a discrepancy in the values. I've added a quote to show the difference.
```
singalsu@piikkisika:~$ sof-eqctl -n 22 -s eq/eq_iir_bassboost.txt
Control size is 304.
Applying configuration "eq/eq_iir_bassboost.txt" into device hw:0 control numid=22.
4607827,0,116,"50331648",0,0,0,0,116,2,1,0,0,0,0,0,0,2,2,0,0,0,0,3227172081,2141520527,536653443,3221660410,536653443,0,16384,3260252783,2107733822,161646111,3961037800,172645501,4294967294,27910
Error: failed TLV write.
singalsu@piikkisika:~$ sof-eqctl -n 22
Control size is 304.
Retrieving configuration for device hw:0 control numid=22.
Success.
4607827,0,116,50356224,0,0,0,0,116,2,1,0,0,0,0,0,0,2,2,0,0,0,0,3227172081,2141520527,536653443,3221660410,"536653443",0,16384,3260252783,2107733822,161646111,3961037800,172645501,4294967294,27910
```

ranj063 commented:
@singalsu i think the way this should is that if the SET_DATA ipc in bytes_ext_put() fails, we should not cache the new values so we're aligned with the error reported by sof-eqctl, which is what your PR does.
I think the more efficient way to deal with this problem is to change the reply message for SET_DATA IPC. Today, the reply message also includes the data that was sent to the DSP but we dont need to do that right? All we need to know is if the command was succesful or not. I think there's an issue filed to do this already

ranj063 commented:
@singalsu looking more into this, I think this PR is a good way to solve the problem. We should update the scontrol data iff the IPC succeeds.

ranj063 commented:
@singalsu this patch is good to go after you update your commit title and message.

lgirdwood commented on sound/soc/sof/control.c:
```diff
@@ -431,24 +432,38 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
 
-	if (copy_from_user(cdata->data, tlvd->tlv, header.length))
+	/* Allocate temporary buffer for new cdata */
+	new_cdata = kzalloc(sizeof(*new_cdata) + header.length, GFP_KERNEL);
```
Is header.length validated anywhere as this can potentially allocate 4G and run OOM.

singalsu replied:
Good point, I need to find out a check for it.

singalsu replied:
I could find nothing from ALSA/ASoC headers so I defined SND_SOF_CTL_TLV_LENGTH_MAX in sof-priv.h. It's set now to 1Mbyte. It's a lot but should avoid maintaining the value up for now. FW will  anyway error if allocation there exceeds available and fail the IPC cleanly.

singalsu commented:
I've now addressed your comments, can you please check again this!

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -56,6 +56,9 @@
 	SNDRV_PCM_FMTBIT_S32_LE | SNDRV_PCM_FMTBIT_FLOAT)
 
+/* Set upper limit for TLV bytes length to 1 MB */
+#define SND_SOF_CTL_TLV_LENGTH_MAX 1048576
```
(1024*1024) or something easier to read.

singalsu replied:
ok

singalsu replied:
Done.

plbossart commented on sound/soc/sof/control.c:
```diff
@@ -416,6 +417,14 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
+				    header.length, SND_SOF_CTL_TLV_LENGTH_MAX);
+		return -EINVAL;
+	}
+
```
can we put this check in a different patch?

kv2019i replied:
This check is not that meaningful in a separate patch. If we put this in a separate patch, I'd also use SND_SOF_CTL_TLV_LENGTH_MAX in topology.c:sof_control_load_bytes() where kzalloc() is now done based on value read from user-space.

singalsu replied:
OK, I'll add the same check there as well and separate it into first commit and leave the actual feature fix related code to second commit.

singalsu replied:
Done!

plbossart commented on sound/soc/sof/control.c:
```diff
@@ -405,6 +405,7 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
 	struct snd_ctl_tlv header;
```
Commit message is rather unclear:

This patch avoids the scontrol persistent storage to be written
over by firmware fail response if the IPC send was not
successful. 

What? Did you mean 
"with the existing code, the configuration is saved in kcontrol cached data even when an IPC fails".


Such situation is possible e.g. if processing
component like EQ responds to IPC with -EBUSY when new
configuration data is passed from user space. The firmware reply
in case of error was copied directly to scontrol data in function
tx_wait_done() and the fail response when re-applied by PM
restore caused another IPC fail.

That last part is unclear. Did you mean "the previous configuration that was saved after the previous IPC failure will be applied by PM restore"? That would be fine, but what is the issue. Why would this cause another IPC to fail? The fail is not a result of the configuration proper but the state of the DSP when the configuration was passed. There is something really fishy about the explanations, sorry.

ranj063 replied:
@plbossart @singalsu I think the problem is that when the IPC fails with -EBUSY, the firmware does not write the control data into the mailbox. So what is copied into the control data in the driver when the IPC reply is received is invalid data. This is what causes the very next IPC to the comp to fail during PM restore.

singalsu replied:
I'll improve the description text as well. The sequence of error is a bit complicated. I didn't first realize that the data to stored scontrol can loop via firmware. And that the response will be re-applied as IPC later. The second IPC fail is not even relevant, I'll mention only the first issue.

singalsu replied:
Done, hope it's better now!

kv2019i commented on sound/soc/sof/control.c:
```diff
@@ -431,24 +440,38 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
 				    "error: Wrong ABI magic 0x%08x.\n",
-				    cdata->data->magic);
+				    new_cdata->data->magic);
+		kfree(new_cdata);
```
There starts to be so many many exit points, that adding a label at end of function and goto to error starts to be more readable. But I'm ok with this as well.

singalsu replied:
I don't myself like "goto". As long as OK with reviewers I'd prefer this way (just one line exit task).

singalsu commented:
@plbossart @kv2019i Can you please review this again. I've now addressed your comments (hope so!).

kv2019i commented on sound/soc/sof/control.c:
```diff
@@ -439,24 +440,38 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
+		dev_err_ratelimited(sdev->dev,
+				    "error: Incompatible ABI version 0x%08x.\n",
+				    new_cdata->data->abi);
+		kfree(new_cdata);
```
Was about to comment that why modify 80+ lines but oh wait, you are modifying s/cdata/new_cdata/ ... -> Ok

singalsu commented:
@plbossart The invalid data was returned by DSP. This is also related to the issue we entered that IPC replies the same size as request are longer than needed. I'll re-find the code that does it.

singalsu commented:
@plbossart The code that leaves/places the invalid IPC data to buffer is here in 898 onwards in firmware IPC handler.c:

```
	/* get component values */
	ret = ipc_comp_cmd(comp_dev->cd, cmd, _data, SOF_IPC_MSG_MAX_SIZE);
	if (ret < 0) {
		trace_ipc_error("ipc: comp %d cmd %u failed %d", data.comp_id,
				data.cmd, ret);
		return ret;
	}

	/* write component values to the outbox */
	if (_data->rhdr.hdr.size <= MAILBOX_HOSTBOX_SIZE &&
	    _data->rhdr.hdr.size <= SOF_IPC_MSG_MAX_SIZE) {
		mailbox_hostbox_write(0, _data, data.rhdr.hdr.size);
		ret = 1;
	} else {
		trace_ipc_error("ipc: comp %d cmd %u returned %d bytes max %d",
				data.comp_id, data.cmd, _data->rhdr.hdr.size,
				MIN(MAILBOX_HOSTBOX_SIZE,
				    SOF_IPC_MSG_MAX_SIZE));
		ret = -EINVAL;
	}

	return ret;

```
I think the data that is received by tx_wait_done() is some old outbox content. It failed in next IPC to ABI header check (missing totally). I wonder if the error handling in FW side should write a data block with receive header size set to zero in addition to just showing "ipc: comp 2 cmd 3 failed -16".

But anyway such response would invalidate the data for PM restore so I think my approach is correct and needed.

Edit: Adding that I will improve the commit message to have explanation.

singalsu commented:
I updated this PR. I added sentence _"The firmware outbox is not updated by the FW side IPC framework when a component command returns an error."_ Not sure if it's enough but it's better than previous.

@lgirdwood @tlauda @plbossart @keyonjie @ranj063 I wonder if the FW should respond to IPC fails with message with data.rhdr.hdr.size = 0 if there's nothing else to report back. Is the current behavior a bug in FW side? Could it break something else than just this case?

ranj063 commented:
> Sorry @singalsu this commit message is just not going to cut it.
> 
> "This patch prevents the scontrol persistent storage from being
> updated with invalid data (in tx_wait_done() function) if the IPC
> send was not successful. The invalid data is seen in the IPC
> buffer if DSP responds with -EBUSY or other error. The firmware
> outbox is not updated by the FW side IPC framework when a component
> command returns an error."
> 
> You have to explain why on earth the data passed by the user and transmitted to the firmware would then be returned by the firmware to be saved into a cached area. Why not save the data passed by the user directly into a cached area _before_ passing it over IPC? Does the firmware change the configuration data and why? Also it's not useful to add unnecessary details such as firmware outbox or tx_wait_done.

@plbossart this is a flaw in the current flow. There are 2 reasons why this happens:
1. The message and the reply pointers in the IPC point to the same control_data buffer today. And thats the reason the data returned from the DSP is being copied into the control data. So copying back the data from the DSP was unintentional. It happens because the pointers are the same. This is what Seppo is fixing in this PR
2. But fundamentally, the problem is because we send the data to the DSP and expect the data back from the DSP, which is totally unnecessary. We should fix that too. All that the DSP should return is whether the data sent was applied successfully or not.

singalsu commented:
@ranj063 Thanks for clear explanation. I tried to say also that but I wasn't successful.

So how do we proceed with this? Reword the commit message or change more the IPC flow? I can update the wording meanwhile but we can close this PR if we go for the larger latter option.

plbossart commented:
> > Sorry @singalsu this commit message is just not going to cut it.
> > "This patch prevents the scontrol persistent storage from being
> > updated with invalid data (in tx_wait_done() function) if the IPC
> > send was not successful. The invalid data is seen in the IPC
> > buffer if DSP responds with -EBUSY or other error. The firmware
> > outbox is not updated by the FW side IPC framework when a component
> > command returns an error."
> > You have to explain why on earth the data passed by the user and transmitted to the firmware would then be returned by the firmware to be saved into a cached area. Why not save the data passed by the user directly into a cached area _before_ passing it over IPC? Does the firmware change the configuration data and why? Also it's not useful to add unnecessary details such as firmware outbox or tx_wait_done.
> 
> @plbossart this is a flaw in the current flow. There are 2 reasons why this happens:
> 
> 1. The message and the reply pointers in the IPC point to the same control_data buffer today. And thats the reason the data returned from the DSP is being copied into the control data. So copying back the data from the DSP was unintentional. It happens because the pointers are the same. This is what Seppo is fixing in this PR
> 2. But fundamentally, the problem is because we send the data to the DSP and expect the data back from the DSP, which is totally unnecessary. We should fix that too. All that the DSP should return is whether the data sent was applied successfully or not.

Well, duh, if the IPC does things that are not necessary, which creates additional errors then let's fix it instead of applying a work-around.

ranj063 commented:
> @ranj063 Thanks for clear explanation. I tried to say also that but I wasn't successful.
> 
> So how do we proceed with this? Reword the commit message or change more the IPC flow? I can update the wording meanwhile but we can close this PR if we go for the larger latter option.

@singalsu the change to not send the data back from the DSP is quite simple too. We just need to change the reply buffer when sending the IPC. If I am not wrong, this should not need any FW change as well. But we'll still need a part of the changes you have in this PR. Let me know if you want me to take it up and implement that part.

singalsu commented:
@plbossart OK I will rename this PR and use only the topology kzalloc check and change of return code passing parts.

@ranj063 I can do the other fix into another PR. I'll ask help if I would get stuck but at least now it looks like a simple change.

singalsu commented:
@ranj063 The fix will be in snd_sof_ipc_set_get_comp_data(). I think I'll allocate a temp buffer for reply_data. In case of "set" I will discard the reply. In case of "get" I will copy the response to scontrol. Does this sound OK to you?

singalsu commented:
@ranj063 @plbossart @kv2019i I did the above change of separating IPC request and response buffers and it "works" now. I didn't yet update this PR.

However with this there's the same impact that I demonstrated above with just small tx_wait_done() edit: The sof-eqctl works differently depending on PM suspend enabled vs. disabled. PM restore will re-apply the attempted ext bytes control in next idle->active transition when next playback starts. While with suspend disabled the new coefficients won't be applied.

To prevent the suspend enabled/disable state to impact the operation I need to use the temporary buffer in ext-bytes send like I did in this last PR version.

Do you want to me include to PR also the work of separating IPC request and response flows? EQ control works without, but there might be other things that it might fix. Also I hope it causes no new bugs, or in other words nothing relies on that firmware loop for scontrols data (if firmware modified the data).

plbossart commented:
> @ranj063 @plbossart @kv2019i I did the above change of separating IPC request and response buffers and it "works" now. I didn't yet update this PR.
> 
> However with this there's the same impact that I demonstrated above with just small tx_wait_done() edit: The sof-eqctl works differently depending on PM suspend enabled vs. disabled. PM restore will re-apply the attempted ext bytes control in next idle->active transition when next playback starts. While with suspend disabled the new coefficients won't be applied.

Huh? if you don't suspend, the coefficients are still valid in the DSP memory so why would you reapply them? What is the expectation and the issue? There must be a side effect where the EQ state is lost, but it can't be due to the suspend/resume state machine.

singalsu commented:
> > @ranj063 @plbossart @kv2019i I did the above change of separating IPC request and response buffers and it "works" now. I didn't yet update this PR.
> > However with this there's the same impact that I demonstrated above with just small tx_wait_done() edit: The sof-eqctl works differently depending on PM suspend enabled vs. disabled. PM restore will re-apply the attempted ext bytes control in next idle->active transition when next playback starts. While with suspend disabled the new coefficients won't be applied.
> 
> Huh? if you don't suspend, the coefficients are still valid in the DSP memory so why would you reapply them? What is the expectation and the issue? There must be a side effect where the EQ state is lost, but it can't be due to the suspend/resume state machine.

My point is that having suspend enabled / suspend disabled should not impact what response EQ is set for. The behavior should not be impacted by this choice. PM restore re-applies what's in cache (if the error was -EBUSY earlier it can be successful later without user getting feedback). The TLV bytes put function places the request into cache prior to sending it. So my view is that the request can be put to cache only after it has been accepted like I do in this PR. 

The separation of flows is another matter. Do you want me to include it?

plbossart commented:
> My point is that having suspend enabled / suspend disabled should not 
> impact what response EQ is set for. The behavior should not be impacted 
> by this choice. PM restore re-applies what's in cache (if the error was 
> -EBUSY earlier it can be successful later without user getting 
> feedback). The TLV bytes put function places the request into cache 
> prior to sending it. So my view is that the request can be put to cache 
> only after it has been accepted like I do in this PR.

Why? We should send the data to the kernel, save it, and then send it to 
the DSP. There is no reason why we should wait for the results of the 
IPC, is there?

ranj063 commented:
> Why? We should send the data to the kernel, save it, and then send it to the DSP. There is no reason why we should wait for the results of the IPC, is there?

@plbossart @singalsu I see both your points. Seppo's point is that when the IPC returns an -EBUSY, the user sees an error. So the assumption is that the new config was not applied. So the kernel should not save the new config to not confuse the user.

Pierre's point is that we should save the new data anyway and let PM apply it at next restore. In this case, what we return from the put() IO matters. It should not return an errro

plbossart commented:
> > Why? We should send the data to the kernel, save it, and then send it to the DSP. There is no reason why we should wait for the results of the IPC, is there?
> 
> @plbossart @singalsu I see both your points. Seppo's point is that when the IPC returns an -EBUSY, the user sees an error. So the assumption is that the new config was not applied. So the kernel should not save the new config to not confuse the user.
> 
> Pierre's point is that we should save the new data anyway and let PM apply it at next restore. In this case, what we return from the put() IO matters. It should not return an errro

Well, this case is just part of the more general use case where the application writes kcontrol values, and we don't have any similar case where we depend on the IPC results with the other types of kcontrols. So either we avoid saving the settings in kernel cached values for everyone, or we save it for all. I find it odd to define a specific behavior only for a type of kcontrol.

singalsu commented:
Seems I'm pushing too much this, I'll give up to proceed with other things :^). I understand your view to have symmetry with other controls. With the IPC flows separated approach I no more get PM cache corruption and it's the most important thing since it fixes issue #753. I'll update this PR with that work.

I think that if DSP doesn't accept the bytes control the user should be returned an error. Though the user application should be more verbose than today to differentiate -EINVAL (e.g. corrupted configuration) and -EBUSY. I can change sof-eqctl to print a warning instead for that case if possible.

The different operation with PM suspend on vs. off remains but it's not a big issue.

Edit: Despite seeing it "works" I still need to test more the new approach more since the change is larger and impacts all IPC. ETA is tomorrow!

singalsu commented:
I just updated this PR. Please review carefully the IPC impact since this impacts now practically every component's IPC.

lyakh commented on sound/soc/sof/control.c:
```diff
@@ -416,6 +416,14 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
+		dev_err_ratelimited(sdev->dev,
+				    "error: Bytes size %d exceeds max %d.\n",
+				    header.length, SND_SOF_CTL_TLV_LENGTH_MAX);
+		return -EINVAL;
```
I'm not 100% sure but for this kind of error it might be better to use ENOBUFS, ENOSPC or ENOMEM seem to be less appropriate but might be an option too. Same below.

singalsu replied:
Yep, thanks for suggestion.

kv2019i replied:
@singalsu @lyakh I was wondering the same as well, but I think EINVAL is more appropriate as this is exceeding a limit that is part of the interface. ENOMEM/NOSPC are typically for dynamic allocation where the state of the system is such that resource allocation fails. In this case this is a programming error so I think EINVAL is most appropriate.

lyakh commented on sound/soc/sof/control.c:
```diff
@@ -416,6 +416,14 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
+	if (header.length > SND_SOF_CTL_TLV_LENGTH_MAX) {
+		dev_err_ratelimited(sdev->dev,
+				    "error: Bytes size %d exceeds max %d.\n",
+				    header.length, SND_SOF_CTL_TLV_LENGTH_MAX);
```
header.length is unsigned, too, I think the comparison will be unsigned too, so I'd use "%u" for printing header.length as well. Same below

singalsu replied:
ok

singalsu commented:
I just fixed the >80 chars code line in this PR. There are no actual code changes.

lyakh commented on sound/soc/sof/ipc.c:
```diff
@@ -696,14 +697,27 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_ipc *ipc,
 			dev_err(sdev->dev, "error: set/get ctrl ipc comp %d\n",
 				cdata->comp_id);
+		else
+			if (!send)
```
Please, put "else" and "if" on the same line, you'll save one level of indentation that way.

singalsu replied:
Done, good to know this. Line lengths are a pain so good to get some help from coding style.

plbossart commented on sound/soc/sof/control.c:
```diff
@@ -471,11 +471,11 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
+					    SOF_IPC_COMP_SET_DATA,
+					    SOF_CTRL_TYPE_DATA_SET,
+					    scontrol->cmd,
+					    true);
```
@juimonen @singalsu I think this fixes an error rather than change the functionality. I don't think we ever expected to return the pm status here?
@ranj063 wadyathink?

plbossart replied:
also the commit message should be ASOC: SOF: IPC: fix bytes_ext_put() return status

ranj063 replied:
@plbossart yes this was an oversight earlier

singalsu replied:
The commit impacts several functions so I kept the commit title more generic. Is it OK?

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -696,14 +697,25 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_ipc *ipc,
+		else if (!send)
+			memcpy(cdata, reply_cdata, reply_cdata->rhdr.hdr.size);
 
+		kfree(reply_cdata);
```
I have absolutely zero understanding of this. I've spent 15 mn looking at the code and I guess there are assumptions I am not aware of in the IPC flow.
Someone else needs to review this, I have zero context and can't spend more time on this.

ranj063 replied:
@plbossart I have discussed the next steps for this PR with @singalsu . Basically, the direction we've taken to align all kcontrol IO to check for the status of the IPC response. The kcontrol data cache should be updated only upon a successful IPC. Seppo will work on it next week.

singalsu replied:
@ranj063 It's not yet ready for PR but I have re-worked functions snd_sof_volume_put(), snd_sof_switch_put(), snd_sof_enum_put(), snd_sof_bytes_put(), and snd_sof_bytes_ext_put(). The IPC is first sent via a temp buffer, and copied to scontrol if return value was zero.

I tried that I still can update EQs, change volume. I need to see what else I can test with our topologies.

singalsu commented:
I just updated this PR per previous discussion. The change is very invasive so it needs a thorough review and tests. I have been able to test only so far volume put and ext bytes but. Switch, enum, and bytes put has not been tested. I wonder what the CI tests coverage is for these.

singalsu commented:
@plbossart What do I need to do to get the "unclear" status cleared? 

(also this PR needs a rebase due to conflict, I'll do that tomorrow)

plbossart commented on sound/soc/sof/control.c:
```diff
@@ -104,19 +104,19 @@ int snd_sof_volume_put(struct snd_kcontrol *kcontrol,
 	}
```
The commit message is convoluted: 
The return value is passed from firmware IPC instead of pm function success value.

I think you meant "The existing code returns the status of the pm_runtime function but we should return the status of the IPC"

singalsu replied:
Yep, That's much better. Thanks I'll use that.

plbossart commented on sound/soc/sof/control.c:
```diff
@@ -82,8 +82,10 @@ int snd_sof_volume_put(struct snd_kcontrol *kcontrol,
 		(struct soc_mixer_control *)kcontrol->private_value;
```
generic comment: I think we are confusing two layers.
a) the IPC failed due to scheduling/timeout issues
b) the data provided to the DSP is invalid.

This PR does not explain how these two different cases are handled. 
In case a) it's to be hoped that the IPC timeout will not happen again on a resume, and this was a transient error.
In case b) the setting will never work unless the application changes the values. It's not a matter of IPC proper but of bad data being rejected.

And I am personally not aware of any specific error-handling policy for controls, not sure either how the two cases should be handled.

singalsu replied:
Currently there's no separated handling of -EBUSY (The request is valid but DSP cannot do it now), -ENOMEM (DSP cannot allocate now the RAM, but in theory could have success another time), and -EINVAL (illegal request, will be always rejected). All of these are treated as error. I left with this PR the behavior as is. It would be a risk to place the request into scontrols cache for state restore. It's quite possible that they would fail in next attempt too by PM restore. Also then it wouldn't be no more just PM restore but a delayed controls apply feature.

Without more detailed information about errors and by keeping PM restore as just restore it's better to keep previous accepted and applied request in cache. This PR implements protect for illegal request to overwrite a previous good request in PM cache.

The DSP responds in addition to 0/-EINVAL/... etc. with random data words. I think it should instead respond with an empty message if there is no additional data to respond with. But since kernel needs to be compatible with previous ABI versions the kernel needs to be made immune against corrupt data from DSP. This patch implements that by not letting DSP response to be written over to scontrols cache. The temporary allocated buffer content is discarded when it's freed.

So, I don't think I need to change the code based on how I interpret your feedback. I'll rebase today and improve commit text. Commit texts are tricky so please help me to get it right to something that works for next level reviews.

singalsu replied:
@plbossart @ranj063 The previous PR version implemented instead in lower level IPC code separation of request and response buffers. This again doesn't use it. Both protecting the scontrols cache and separating the flows could be done if you prefer. Though with current IPC needs it does not seem to be necessary (dual kzallocs and data copies done). Please let me know your preference for that as well to get this PR technically into best possible shape.

plbossart commented:
> Currently there's no separated handling of -EBUSY (The request is valid 
> but DSP cannot do it now), -ENOMEM (DSP cannot allocate now the RAM, but 
> in theory could have success another time), and -EINVAL (illegal 
> request, will be always rejected). All of these are treated as error. I 
> left with this PR the behavior as is.

That part is fine, but your commit message needs to explain all this.

> The DSP responds in addition to 0/-EINVAL/... etc. with random data 
> words. I think it should instead respond with an empty message if there 
> is no additional data to respond with. But since kernel needs to be 
> compatible with previous ABI versions the kernel needs to be made immune 
> against corrupt data from DSP. This patch implements that by no letting 
> DSP response to be written over to scontrols cache.

That part is not clear at all. Again I don't see why we use the response 
and write it in the scontrols cache. I don't even understand why we need 
more than an error code returned by the DSP?

plbossart commented:
> The previous PR version implemented also in 
> lower level IPC code separation of request and response buffers. This 
> again doesn't use it. Both protecting the scontrols cache and separating 
> the flows could be done if you prefer. Though with current IPC needs it 
> does not seem to be necessary (dual kzallocs and data copies done). 
> Please let me know your preference for that as well to get this PR 
> technically into best possible shape.

I don't have a clue, I don't understand why you need the response buffer 
for a _put.

singalsu commented:
> Currently there's no separated handling of -EBUSY (The request is valid but DSP cannot do it now), -ENOMEM (DSP cannot allocate now the RAM, but in theory could have success another time), and -EINVAL (illegal request, will be always rejected). All of these are treated as error. I left with this PR the behavior as is.
> **That part is fine, but your commit message needs to explain all this.**

OK got it.

> The DSP responds in addition to 0/-EINVAL/... etc. with random data words. I think it should instead respond with an empty message if there is no additional data to respond with. But since kernel needs to be compatible with previous ABI versions the kernel needs to be made immune against corrupt data from DSP. This patch implements that by no letting DSP response to be written over to scontrols cache.
> **That part is not clear at all. Again I don't see why we use the response and write it in the scontrols cache. I don't even understand why we need more than an error code returned by the DSP?**

I don't think the firmware response needs to be a copy of the request. I don't know why it is like it's today (return the "set" request data if it was successful, return random bytes data if fail).  Just the return code would be sufficient for "set". Do you know @lgirdwood ?  This patch omits the response unless the request is a "get". Also the current data flow from "set" request -> response -> scontrols cache is questionable. Do we need for the firmware possibility to alter the request that PM restore would apply later? I'm not aware of such need. Also, this PR prevents such scenario.

Can the people who originally designed the IPC help me with the direction. My motivation is to get sof-eqctl to work reliably and correctly indicate the user whether the control was applied or not. This thread is expanding to overall IPC cleanup. It's OK but I'm missing information why things are done as they are today.

singalsu commented:
> The previous PR version implemented also in lower level IPC code separation of request and response buffers. This again doesn't use it. Both protecting the scontrols cache and separating the flows could be done if you prefer. Though with current IPC needs it does not seem to be necessary (dual kzallocs and data copies done). Please let me know your preference for that as well to get this PR technically into best possible shape.
> **I don't have a clue, I don't understand why you need the response buffer for a _put.**

Me neither, but it's how it works today. Can you help @lgirdwood ?

singalsu commented:
I just pushed a new version of this PR. The only code change is a minor merge conflict fix. I've improved the commit texts for the two commits those were not clear enough.

I think I've fully addressed only one of three issues raised by @plbossart (just commit text improve). For doing more actual code changes related to the other two points above need help from e.g. @lgirdwood to understand why the request/response IPC is like it is. I can add more fixes to this patch set but I need guidance about desired IPC flow and how to address previous ABI version if we need to bump the ABI.

singalsu commented:
@ranj063 @plbossart @lgirdwood Can you please help me proceed with this. Is this the preferred approach or do you want me to start a larger IPC cleanup effort starting from FW?

keyonjie commented on sound/soc/sof/control.c:
```diff
@@ -110,6 +121,15 @@ int snd_sof_volume_put(struct snd_kcontrol *kcontrol,
+	 */
+	scontrol->control_data = pcdata;
+	if (!ret)
+		memcpy(pcdata->chanv, cdata->chanv, chanv_size);
```
@singalsu This looks somewhat convoluted to me. I suppose we have bot .put() and .get(), for each .put(), we just send the target values to FW directly, and then we will use .get() to get the final values from FW directly and update it to application user, so cache here actually no needed?

Even the .put() setting fails in FW, the FW itself must using some other real values, using .get() to get the real values make sense?

singalsu replied:
> @singalsu This looks somewhat convoluted to me. 

I didn't change IPC principle with this PR just fixed the scontrols cache corrupt bug with ext bytes put() and a hard to predict impact to FW configuration with get(). Those make use of sof-eqctl tool strange and likely also other user space originated control. Then the same fixed/improved IPC flow was extended to other controls. I tried to get feedback if this approach is OK or does it need a larger redesign. It's not cleanest possible I agree.

The kernel request/response buffers could be separate and the FW should not respond with random bytes the size of request if there's a fail. But that would be larger change.

>I suppose we have bot .put() and .get(), for each .put(), we just send the target values to FW directly, and then we will use .get() to get the final values from FW directly and update it to application user, so cache here actually no needed?

The scontrols cache is solely for PM restore to work. It uses for FW configuration restore the last put() or get() data. The PM suspend is not doing any context store for audio components. If there would be no PM suspending the scontrols cache would not be needed.

> 
> Even the .put() setting fails in FW, the FW itself must using some other real values, using .get() to get the real values make sense?

This code does not do get() after a failed put(). The assumption is that the last successful put() command can be use to restore firmware configuration.

Did this address your concern?

singalsu replied:
@keyonjie ?

singalsu commented:
> @singalsu I'm see that you have made all requested changes. IIUC you are only caching values on put() and only if the IPC succeeds ? 

@lgirdwood Thanks for review! Yes this patch changes the put() caching to be done after IPC success. 

The get() commands work as earlier and the result goes to cache always. I didn't touch it because I haven't had issues with it.

> Are you querying the usage of the IPC return data and whether this should be cached ?

Sorry, I'm not sure what you meant with this question. 

Concerning usage -- earlier version of this patch was only for ext bytes put() but based on discussion with @plbossart and @ranj063 this last version extended to all put() controls for similar IPC flow for all even if there's today no known issues with them.

Concerning return data -- no, the code does not try to be intelligent with data/status. If status is 0 the data is stored to cache and that's all.

Also this version of patch does not separate the shared request/response buffer so the data flows to cache through FW. In theory FW could alter it. I don't know if it's a good desired feature or not. My previous version of this PR included separation as core of this fix but I interpreted the review that such is not needed or that patch was too invasive (impacted every IPC) since I did it at a lower layer of IPC compared to this version.

singalsu commented:
This PR needs a re-base, I'll do it next.

ranj063 commented:
@singalsu this PR looks good but I guess the biggest open left is why do we need to return the data that is sent to the FW to be returned back to the host? IMO, all we need to reply is acknowledge the receipt/processing of the data and not return it back to the host.

singalsu commented:
@ranj063 Thanks. I don't think we need the put response data to host. I've tried to ask and none of review participants have indicated such need. The FW side should be enhanced to not send unnecessary response data (should I do the PR?) and the kernel side should ignore it for "put" if such happens. Latter is not fully implemented into this PR. This patch does not prevent firmware to overwrite and modify the request data that goes to cache like a previous PR version did. Should I add it back into the lower level tx_wait_done() code or is this level of security sufficient?   

I remember we've talked about restoring EQ state if we had also support for switching the EQ presets for channels but currently the driver doesn't support it (the code exists in FW but it could be removed if we decide to drop the feature). This IPC feature we now have would allow the cache for PM restore to track the run-time made updates for EQ configuration. But I doubt the IPC has been been designed to support that purpose. I'm guessing it's a non-intended feature or been designed to save a bit of RAM. Also I don't think such bit more advanded context save for PM is a replacement of true context save if such would be needed.

ranj063 commented:
> @ranj063 Thanks. I don't think we need the put response data to host. I've tried to ask and none of review participants have indicated such need. The FW side should be enhanced to not send unnecessary response data (should I do the PR?) and the kernel side should ignore it for "put" if such happens. Latter is not fully implemented into this PR. This patch does not prevent firmware to overwrite and modify the request data that goes to cache like a previous PR version did. Should I add it back into the lower level tx_wait_done() code or is this level of security sufficient?

@singalsu I think adding it to tx_wait_done might affect the IPC flow overall. SO better to just do it in the kcontrol IO callbacks. And yes, could you also submit the firmware PR?

singalsu commented:
> I think adding it to tx_wait_done might affect the IPC flow overall. SO better to just do it in the kcontrol IO callbacks. 

Yes I recall we discussed for that PR version that it was too invasive.

> And yes, could you also submit the firmware PR?

Sure.

singalsu commented:
Oh seems like this issue got somewhat different with commit 38d0e9fc227c7876d09754863adc88aeca6dd205 into control.c. I think it is needed to accept that if user space passes invalid control to cache it will remain there and continue to fail every time firmware become active. It's fundamental to this commit that DSP is not waked up for control IPC so we no more can test the result before updating cache.

Instead I'll now focus to IPC flow cleanup and prevent corrupt response from a temporary error (-EBUSY) to be circulated to cache.

singalsu commented:
I'll close this to avoid cluttering the list of pull requests. since I won't have a solution within one month (other work plus my vacation)  to propose for Linux issue #753. I'll re-open this after I have it.
